go/service/gregor.go

```diff
@@ -194,24 +193,52 @@ func newGregorHandler(g *globals.Context) *gregorHandler {
 		firstConnect:    true,
 		pushStateFilter: func(m gregor.Message) bool { return true },
 		badger:          nil,
-		chatHandler:     chat.NewPushHandler(g),
 		broadcastCh:     make(chan gregor1.Message, 10000),
 	}
+	return gh
+}
 
+// Init starts all the background services for managing connection to Gregor
+func (g *gregorHandler) Init() {
 	// Attempt to create a gregor client initially, if we are not logged in
 	// or don't have user/device info in G, then this won't work
-	if err := gh.resetGregorClient(); err != nil {
-		g.Log.Warning("unable to create push service client: %s", err)
+	if err := g.resetGregorClient(); err != nil {
+		g.Warning(context.Background(), "unable to create push service client: %s", err.Error())
 	}
 
 	// Start broadcast handler goroutine
-	go gh.broadcastMessageHandler()
+	go g.broadcastMessageHandler()
 
-	return gh
+	// Start the app state monitor thread
+	go g.monitorAppState()
 }
 
-func (g *gregorHandler) GetClient() rpc.GenericClient {
-	return g.cli
+func (g *gregorHandler) monitorAppState() {
+	// Wait for state updates and react accordingly
+	for {
+		state := <-g.G().AppState.NextUpdate()
+		switch state {
+		case keybase1.AppState_BACKGROUNDACTIVE:
+			fallthrough
+		case keybase1.AppState_FOREGROUND:
+			g.Shutdown()
+			g.chatLog.Debug(context.Background(), "foregrounded, reconnecting")
+			if err := g.Connect(g.uri); err != nil {
+				g.chatLog.Debug(context.Background(), "error reconnecting")
+			}
+		case keybase1.AppState_INACTIVE:
+			g.chatLog.Debug(context.Background(), "backgrounded, shutting down connection")
+			g.Shutdown()
+		}
```

Comment:
No, you always get `INACTIVE` before that, which for managing the connection is all you need.